IFC4x3/Sections/Core data schemas/Schemas/IfcProductExtension/Functions/IfcAlignmentRepresentation/Expression.txt

```diff
@@ -0,0 +1,42 @@
+(Alignment : IfcAlignment) : IfcBoolean;
+	
+	LOCAL
+		AlignmentRepresentations : SET OF IfcShapeRepresentation;
+		TwoDRep : SET OF IfcShapeRepresentation;
+		ThreeDRep : SET OF IfcShapeRepresentation;
+	END_LOCAL;
+
+	(* no representation is okay *)
+	IF NOT EXISTS(Alignment.Representation) THEN
+		RETURN(TRUE);
+	END_IF;
+	
+	(* Alignment.Representation exists so assign *)
+	AlignmentRepresentations := Alignment.Representation.Representations;
+
+	(* one representation is either Curve2D or 3D *)
+	IF SIZEOF(AlignmentRepresentations) = 1
+	AND
+	(SIZEOF(QUERY(twoDee <* AlignmentRepresentations | twoDee.RepresentationType = 'Curve2D')) = 1
+	OR
+	SIZEOF(QUERY(threeDee <* AlignmentRepresentations | threeDee.RepresentationType = 'Curve3D')) = 1)
+	THEN
+		RETURN(TRUE);
+	END_IF;
+	
+	(* 2 representations and 3D BaseCurve reference equals 2D representation *)
+	IF SIZEOF(AlignmentRepresentations) = 2 THEN
+	
+		TwoDRep := QUERY(twoDee <* AlignmentRepresentations | twoDee.RepresentationType = 'Curve2D');
+		ThreeDRep := QUERY(threeDee <* AlignmentRepresentations | threeDee.RepresentationType = 'Curve3D');
+		IF SIZEOF(ThreeDRep) = 1 AND ('IFCGEOMETRYRESOURCE.IFCGRADIENTCURVE' IN TYPEOF(ThreeDRep[1].Items[1])) THEN
+			IF SIZEOF(TwoDRep) = 1 AND ('IFCGEOMETRYRESOURCE.IFCCOMPOSITECURVE' IN TYPEOF(TwoDRep[1].Items[1])) THEN
+				IF ThreeDRep[1].Items[1]\IfcGradientCurve.BaseCurve :=: TwoDRep[1].Items[1] THEN
```

Comment:
What if there are multiple `.Items`?

Reply:
How about a check before:
```
IF SIZEOF(ThreeDRep[1].Items) > 1 THEN RETURN(FALSE);

```
?

Reply:
Another option would be to query and save as local variable:

> curveThreeD := QUERY(itemThreeD <* ThreeDRep[1].Items | 'IFCGEOMETRYRESOURCE.IFCGRADIENTCURVE' in TYPEOF(itemThreeD) );

Similarly for 2D representation (`curveTwoD`).

And then test:

> RETURN SIZEOF(curveThreeD) = 1 AND SIZEOF(curveTwoD) = 1 AND curveThreeD\IfcGradientCurve.BaseCurve :=: curveTwoD;

I'm unsure about the syntax, but hopefully I manage to get the idea across.

Reply:
Mhm, I like that idea. Not sure if I have the time to do it before Thursday though. 😟